pkg/controller/oam/applicationconfiguration/applicationconfiguration.go

```diff
@@ -0,0 +1,315 @@
+// Components and Traits.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	ac := &v1alpha2.ApplicationConfiguration{}
+	if err := r.client.Get(ctx, req.NamespacedName, ac); err != nil {
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetAppConfig)
+	}
+
+	log = log.WithValues("uid", ac.GetUID(), "version", ac.GetResourceVersion())
+
+	workloads, err := r.components.Render(ctx, ac)
+	if err != nil {
+		log.Debug("Cannot render components", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(ac, event.Warning(reasonCannotRenderComponents, err))
+		ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errRenderComponents)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+	}
+	log.Debug("Successfully rendered components", "workloads", len(workloads))
+	r.record.Event(ac, event.Normal(reasonRenderComponents, "Successfully rendered components", "workloads", strconv.Itoa(len(workloads))))
+
+	if err := r.workloads.Apply(ctx, workloads); err != nil {
+		log.Debug("Cannot apply components", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(ac, event.Warning(reasonCannotApplyComponents, err))
+		ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errApplyComponents)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+	}
+	log.Debug("Successfully applied components", "workloads", len(workloads))
+	r.record.Event(ac, event.Normal(reasonApplyComponents, "Successfully applied components", "workloads", strconv.Itoa(len(workloads))))
+
+	// Kubernetes garbage collection will (by default) reap workloads and traits
+	// when the appconfig that controls them (in the controller reference sense)
+	// is deleted. Here we cover the case in which a component or one of its
+	// traits is removed from an extant appconfig.
+	for _, e := range r.gc.Eligible(ac.GetNamespace(), ac.Status.Workloads, workloads) {
+		// https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable
+		e := e
+
+		log := log.WithValues("kind", e.GetKind(), "name", e.GetName())
+		record := r.record.WithAnnotations("kind", e.GetKind(), "name", e.GetName())
+
+		if err := r.client.Delete(ctx, &e); resource.IgnoreNotFound(err) != nil {
+			log.Debug("Cannot garbage collect component", "error", err, "requeue-after", time.Now().Add(shortWait))
+			record.Event(ac, event.Warning(reasonCannotGGComponents, err))
+			ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errGCComponent)))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
```

Comment:
Just to markdown my questions,  it seems that workloads/traits are already applied to the k8s cluster by now so wouldn't we lose track of them if we return here?